core/src/main/scala/caliban/wrappers/Caching.scala

```diff
@@ -0,0 +1,329 @@
+    }
+
+    implicit val ordering: scala.math.Ordering[CacheScope] = ord.toScala
+
+    val _type = __Type(
+      __TypeKind.ENUM,
+      Some("CacheScope"),
+      Some("The scope of the cache"),
+      enumValues = _ =>
+        Some(
+          List(
+            __EnumValue(
+              "PRIVATE",
+              Some("The cache is private to the user"),
+              deprecationReason = None,
+              isDeprecated = false,
+              directives = None
+            ),
+            __EnumValue(
+              "PUBLIC",
+              Some("The cache is public"),
+              deprecationReason = None,
+              isDeprecated = false,
+              directives = None
+            )
+          )
+        )
+    )
+  }
+
+  case class CacheHint(
+    maxAge: Option[Duration] = None,
+    scope: Option[CacheScope] = None,
+    inheritMaxAge: Boolean = false
+  ) { self =>
+    def toHeaderString: String = {
+      val computedMaxAge = maxAge.fold(0L)(_.toSeconds)
+      val cacheable      = computedMaxAge > 0L
+      if (!cacheable) "no-store"
+      else {
+        val b = new StringBuilder()
+        b.append("max-age=")
+        b.append(computedMaxAge)
+        scope.foreach { s =>
+          b.append(", ")
+          b.append(s.toString.toLowerCase)
+        }
+        b.toString()
+      }
+    }
+  }
+
+  object CacheHint {
+    val default: CacheHint = CacheHint()
+  }
+
+  private class Default[-R1](inner: Ref[CachePolicy] => Wrapper[R1]) extends GraphQLAspect[Nothing, R1] {
+    private val _typeCache = new ConcurrentHashMap[String, Option[CacheHint]]()
+
+    def apply[R <: R1](gql: GraphQL[R]): GraphQL[R] = {
+      val wrapper = EffectfulWrapper(
+        cacheOverride.set(None) *> Ref
+          .make(CachePolicy(CacheHint.default))
+          .map(state => staticWrapper(state) |+| fieldWrapper(state) |+| inner(state))
+      )
+
+      gql
+        .withAdditionalDirectives(cacheDirectives)
+        .withAdditionalTypes(CacheScope._type :: Nil)
+        .withWrapper(wrapper)
+    }
+
+    private[caliban] def cacheHintFromType(typ: __Type): Option[CacheHint] = {
+      val key = typ.name.getOrElse(typ.toString)
+      _typeCache.computeIfAbsent(key, _ => typ.directives.flatMap(extractCacheDirective))
+    }
+
+    private def staticWrapper(state: Ref[CachePolicy]): ValidationWrapper[Any] = new ValidationWrapper[Any] {
+      override def wrap[R](
+        f: Document => ZIO[R, CalibanError.ValidationError, ExecutionRequest]
+      ): Document => ZIO[R, CalibanError.ValidationError, ExecutionRequest] = { (d: Document) =>
+        f(d).flatMap { request =>
+          def loop(policy: CachePolicy, field: Field, parentHint: Option[CacheHint]): CachePolicy = {
+            val fieldHint   = extractCacheDirective(field.directives)
+            val isInherited = fieldHint.exists(_.inheritMaxAge)
+            val typeHint    = field.parentType.flatMap(cacheHintFromType)
```

Comment:
Consider making this into a `def` instead so that it's not executed if not needed